[mlir][ODS] Change `get...Mutable` to return `OpOperand &` for single operands

matthias-springer opened the pull request:
The TableGen code generator now generates C++ code that returns a single `OpOperand &` for `get...Mutable` of operands that are not variadic and not optional. `OpOperand::set`/`assign`  can be used to set a value (same as `MutableOperandRange::assign`). This is safer than `MutableOperandRange` because only single values (and no longer `ValueRange`) can be assigned.

E.g.:
```
// Assignment of multiple values to non-variadic operand.
// Before: Compiles, but produces invalid op.
// After: Compilation error.
extractSliceOp.getSourceMutable().assign({v1, v2});
```

jpienaar commented on mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp:
```diff
@@ -2071,29 +2071,36 @@ void OpEmitter::genNamedOperandSetters() {
-            ", ::mlir::MutableOperandRange::OperandSegment({0}u, *{1})", i,
-            emitHelper.getAttr(operandSegmentAttrName, /*isNamed=*/true));
+    body << "  auto range = getODSOperandIndexAndLength(" << i << ");\n";
+    if (operand.isVariadicOfVariadic() || operand.isVariableLength()) {
```
For variadic of variadic case, could this result in changing number of operands without updating the segment attribute? (Folks ran into a foot gun around this the other day).

matthias-springer replied:
The code inside of the "then" branch is identical to what the original code was coding. So there is no difference in functionality for "variadic of variadic" (or variable-length operands).

matthias-springer replied:
I rewrote the code with an early return instead of multiple `if` checks.

matthias-springer committed:
[mlir][TblGen] `get...Mutable` returns `OpOperand &` for single operands

The TableGen code generator now generates C++ code that returns a single `OpOperand &` for `get...Mutable` of operands that are not variadic and not optional. `OpOperand::set`/`assign`  can be used to set a value (same as `MutableOperandRange::assign`). It is safer than `MutableOperandRange` only single values (and no longer `ValueRange`) can be assigned.

E.g.:
```
// Before: Assign multiple values to non-variadic operand (forbidden, but
//         compiles).
// After: Compilation error.
extractSliceOp.getSourceMutable().assign({v1, v2});
```

BEGIN_PUBLIC
No public commit message needed for presubmit.
END_PUBLIC

jpienaar reviewed (approved):
Looks good in general, looks like it'll avoid a few more footguns. Thanks!
